include/onnxruntime/core/session/onnxruntime_cxx_api.h

```diff
@@ -1508,6 +1511,40 @@ struct KernelContext {
   OrtKernelContext* ctx_;
 };
 
+// Logs a message using the provider Ort::Logger, which must be passed by value.
+// Example:
+//   Ort::Logger logger(ort_logger);
+//   ORT_CXX_LOG(logger, ORT_LOGGING_LEVEL_INFO, "Log a message");
+#define ORT_CXX_LOG(logger, severity, message) \
+  logger.LogMessage(severity, __FILE__, __LINE__, static_cast<const char*>(__FUNCTION__), message)
+
+// Logs a formatted message using the provided Ort::Logger, which must be passed by value.
+// Example:
+//   Ort::Logger logger(ort_logger);
+//   ORT_CXX_LOGF(logger, ORT_LOGGING_LEVEL_INFO, "Log an int: %d", 12);
+#define ORT_CXX_LOGF(logger, severity, ...) \
+  logger.LogFormattedMessage(severity, __FILE__, __LINE__, static_cast<const char*>(__FUNCTION__), __VA_ARGS__)
+
+struct Logger : detail::Base<detail::Unowned<const OrtLogger>> {
+  explicit Logger(std::nullptr_t) {}         ///< Create an empty instance to initialize later
+  explicit Logger(const OrtLogger* logger);  ///< Uses an existing ::OrtLogger instance
+
+  OrtLoggingLevel GetLoggingSeverityLevel() const;  ///< Wraps OrtApi::Logger_GetLoggingSeverityLevel
+
+  ///< Wraps OrtApi::Logger_LogMessage. Typically called via the ORT_CXX_LOG macro.
+  Status LogMessage(OrtLoggingLevel log_severity_level, const char* file_path, int line_number,
+                    const char* func_name, const char* message) const noexcept;
+
+  ///< Prints a formatted message to a buffer and then calls OrtApi::Logger_LogMessage.
+  ///< Typically called via ORT_CXX_LOGF macro.
+  Status LogFormattedMessage(OrtLoggingLevel log_severity_level, const char* file_path, int line_number,
+                             const char* func_name, const char* format, ...) const noexcept;
```

Comment:
I am a little uneasy to expose (...) as a public function in C++.

What if we make public interface like
template<typename ..Args> ...

Then expand it and call (....) which we would make private. Then the overload with va_list and additional C headers is not needed.

Reply:
> What if we make public interface like
template<typename ..Args> ...

I'll give that a shot.

> Also, it is not clear what customer would do with the returned Status.

Yeah, I'm not happy with this. I remember discussing that logging should be noexcept so that we can log from catch clauses or destructors. However, if logging fails for some reason, returning a status is the only way to inspect the failure.

Maybe we should have two LogMsg functions at the C++ API level: one that throws and another that does not throw or return a status at all. Thoughts?

Reply:
I like the idea. We just need to express this properly in a macro and document it. Like don't use this one in destructors.